Sources/NIOSOCKS/Channel Handlers/SOCKSClientHandler.swift

```diff
@@ -0,0 +1,196 @@
+//
+// See LICENSE.txt for license information
+// See CONTRIBUTORS.txt for the list of SwiftNIO project authors
+//
+// SPDX-License-Identifier: Apache-2.0
+//
+//===----------------------------------------------------------------------===//
+
+import NIO
+
+/// Data was unexpectedly written or read before the SOCKS proxy
+/// connection has been fully established.
+public struct ProxyNotEstablished: Error {
+    
+}
+
+/// Connects to a SOCKS server to establish a proxied connection
+/// to a host. This handler should be inserted at the beginning of a
+/// channel's pipeline.
+public class SOCKSClientHandler: ChannelDuplexHandler {
+    
+    public typealias InboundIn = ByteBuffer
+    public typealias InboundOut = ByteBuffer
+    public typealias OutboundIn = ByteBuffer
+    public typealias OutboundOut = ByteBuffer
+    
+    private let authenticationDelegate: SOCKSClientAuthenticationDelegate
+    private let supportedAuthenticationMethods: [AuthenticationMethod]
+    private let targetAddress: AddressType
+    
+    private var state: ClientStateMachine
+    private var buffered: ByteBuffer
+    
+    private var bufferedWrites: [(NIOAny, EventLoopPromise<Void>?)] = []
+    
+    public init(
+        supportedAuthenticationMethods: [AuthenticationMethod],
+        targetAddress: AddressType,
+        authenticationDelegate: SOCKSClientAuthenticationDelegate
+    ) {
+        precondition(supportedAuthenticationMethods.count > 0,
+                     "At least one supported authentication method required.")
+        precondition(supportedAuthenticationMethods.count <= 255,
+                     "There can't be more than 255 authentication methods listed.")
+        self.supportedAuthenticationMethods = supportedAuthenticationMethods
+        self.state = ClientStateMachine()
+        self.buffered = ByteBuffer()
+        self.targetAddress = targetAddress
+        self.authenticationDelegate = authenticationDelegate
+    }
+    
+    public func channelActive(context: ChannelHandlerContext) {
+        self.handleAction(self.state.connectionEstablished(), context: context)
+    }
+    
+    public func handlerAdded(context: ChannelHandlerContext) {
+        if context.channel.isActive {
+            guard self.state.shouldBeginHandshake else {
+                return
+            }
+            self.handleAction(self.state.connectionEstablished(), context: context)
+        }
+    }
+    
+    public func channelRead(context: ChannelHandlerContext, data: NIOAny) {
+        
+        // if we've established the connection then forward on the data
+        guard !self.state.proxyEstablished else {
+            context.fireChannelRead(data)
+            return
+        }
+        
+        var buffer = self.unwrapInboundIn(data)
+        self.buffered.writeBuffer(&buffer)
+        do {
+            let action = try self.state.receiveBuffer(&self.buffered)
+            switch action {
+            case .waitForMoreData:
```

Comment:
Could you expand?

Reply:
This block says "if the action is `.waitForMoreData`, do nothing". Why do we have to do that here instead of in the handler function called for all other values of `action` (which in fact does nothing for the action `.waitForMoreData`)?

Reply:
Ah, understood, I've refactored it